### PR TITLE
Create command to list CQRS endpoints

### PR DESCRIPTION
**Pull request by Amoifr**
| Questions         | Answers
| ------------------| -------------------------------------------------------
| Description?      | Create command to list CQRS endpoints
| Type?             | new feature
| BC breaks?        | no
| Deprecations?    | no
| Fixed ticket?     | no
| How to test?      | execute command
| Possible impacts? | github api is subject to ratios and and this command may get over them

### CODE REVIEW

File: src/App/Command/GithubCQRSEndpointsCommand.php [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * @var Github;
+     */
+    protected $github;
+
+    const NUMBER_OF_ITEMS_PER_PAGE = 100;
+    const ADAPTER_REPOSITORY_PATH = 'src/Adapter/';
+    const COMMAND_HANDLER = 'CommandHandler';
+    const QUERY_HANDLER = 'QueryHandler';
+    const HANDLER_SUFFIX = 'Handler';
```
Comment (Progi1984) [inline src/App/Command/GithubCQRSEndpointsCommand.php:26]: Set the scope

File: src/App/Command/GithubCQRSEndpointsCommand.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
+<?php
+
+namespace Console\App\Command;
```
Comment (Progi1984) [inline src/App/Command/GithubCQRSEndpointsCommand.php:3]:
```suggestion
<?php

declare(strict_types=1);

namespace Console\App\Command;
```

File: src/App/Command/GithubCQRSEndpointsCommand.php [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        $rows[] = new TableSeparator();
+        $files = [];
+        $query = 'filename:*Handler.php+repo:Prestashop/Prestashop';
+
+        $api = new Search($this->github->getClient());
+        $api->setPerPage(self::NUMBER_OF_ITEMS_PER_PAGE);
+        $handlers = $api->code($query);
+        $currentItem = 0;
+        $selectedItems = 0;
+        for ($i = 0; $i < $handlers['total_count'] / self::NUMBER_OF_ITEMS_PER_PAGE; ++$i) {
```
Comment (FabienPapet) [inline src/App/Command/GithubCQRSEndpointsCommand.php:63]: Maybe add an option for `NUMBER_OF_ITEMS_PER_PAGE` and use it with `$input->getOption()`

File: src/App/Command/GithubCQRSEndpointsCommand.php [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+            foreach ($handlers['items'] as $handler) {
+                ++$currentItem;
+                $pathName = pathinfo(substr($handler['path'], strlen(self::ADAPTER_REPOSITORY_PATH)))['dirname'];
+                if (str_starts_with($handler['path'], self::ADAPTER_REPOSITORY_PATH)) {
```
Comment (FabienPapet) [inline src/App/Command/GithubCQRSEndpointsCommand.php:67]: Will this script be run with PHP < 8 ? 

if so `str_starts_with` will not work.

File: src/App/Command/GithubCQRSEndpointsCommand.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                    }
+                    ++$selectedItems;
+                    $files[$pathName][$handler['name']] = [
+                        'id' => $currentItem,
+                        'name' => $handler['name'],
+                    ];
+                }
+            }
+            // it stinks, but needed to not exceed quota on github API (https://docs.github.com/en/rest/rate-limit)
+            sleep(7);
```
Comment (FabienPapet) [inline src/App/Command/GithubCQRSEndpointsCommand.php:79]: Needs to be in constant API_CALL_DELAY amybe ?

File: src/App/Service/Github/Search.php [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+namespace Console\App\Service\Github;
+
+use Github\Api\AbstractApi;
+use Github\Api\AcceptHeaderTrait;
+
+/**
+ * Implement the Search API. To override the default class that not support per_page and page parameters,
+ */
+class Search extends AbstractApi
```
Comment (Progi1984) [inline src/App/Service/Github/Search.php:11]: Why not extends the Search class for KNPLabs package for just overriding the good method ?

(And contribute to KNPLabs package, of course).

**Review by Progi1984 — COMMENTED** [review]
Small feedbacks :)

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 2 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:3]: yes !

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:63]: yes to do only one

File: src/App/Service/Github/Search.php (hunk 6 above)
Comment (Amoifr) [reply src/App/Service/Github/Search.php:11]: So.... I' ve adjusted all method of the class it makes sense to override all the class.
And about contrib to the native package will be a bit more tricky. The maintainer removes the getters for per_page and page param in AbstractApi class. I will have to discuss with him before.

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 3 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:63]: It about internal mechanism not external choice of the user.
In fact the default value is 30 and ask to make too many loops that end with a ration limit override. So this constant is just here to no use a pragmatic value directly without context.

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 4 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:67]: Ho shit !!! that's the reason why I've never used it before !

Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:67]: replaced by a classic strpos === 0 ;-) thx @FabienPapet

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 5 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:79]: you're also right .....I'm a spider-pig !

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 1 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:26]: ok thks

File: src/App/Command/GithubCQRSEndpointsCommand.php (hunk 5 above)
Comment (Amoifr) [reply src/App/Command/GithubCQRSEndpointsCommand.php:79]: I can also walk on the ceiling

**Review by matthieu-rolland — APPROVED** [review]
nice work 👍

**Review by matks — APPROVED** [review]
- feedback from @FabienPapet applied
- approved by @matthieu-rolland 
- tested on localhost

**Pull request merged**